### PR TITLE
tmpfiles: introduce an explicit line flag $ for enabling purge logic …

### PR DESCRIPTION
**Pull request by poettering**
…for a line

Let's make the risk of accidental misuse, and mark lines that shall be covered by --purge with an explicit new flag "$".

See: #33349

### CODE REVIEW

File: NEWS [hunk 1]
```diff
@@ -2,6 +2,15 @@ systemd System and Service Manager
...
 CHANGES WITH 257 in spe:
 
+        Incompatible changes:
+
+        * The --purge switch of systemd-tmpfiles (which was added in v256) has
+          been reworked: it will now only apply to tmpfiles.d/ lines marked
+          with the new "$" flag. This is an incompatible change, and means any
+          tmpfiles.d/ files which shall be used together with --purge need to
+          be updated accordingly. This change has been made to make it harder
+          to accidentally delete too many files when using --purge incorrectly.
```
Comment (YHNdnzj) [inline NEWS:12]: Ugh, all other incompatible changes are listed in the section below already

**Review by bluca — CHANGES_REQUESTED** [review]
I'd much prefer if the logic was inverted - add a specifier that makes the line ignored. That way the few catch-all directories (like /home for btrfs) can be marked, and the rest which is package specific can work out of the box

**Comment by poettering** [comment]
> I'd much prefer if the logic was inverted - add a specifier that makes the line ignored. That way the few catch-all directories (like /home for btrfs) can be marked, and the rest which is package specific can work out of the box

That would make the excercise pointless. The whole idea was to make purge behaviour a per-line opt-in.

File: NEWS (hunk 1 above)
Comment (poettering) [reply NEWS:12]: this one is a compat break that now happened, the other ones are plans for the future

**Comment by bluca** [comment]
I don't think it would be pointless? It would make a lot of sense to me, to be able to guard top level directories with this, for example home.conf, etc.conf, var.conf, and more. All the things that create top level or first level directories, that would have a high impact and that are shared by many different programs/use cases, and it's extremely unlikely to actually be meant to be purged. While still leaving the majority as-is, which is per-package very specific and narrow stuff.

**Comment by poettering** [comment]
Oh man, I thought we had agreement to tighten the rules like this, after the fuck-up people ran into. We talked about this, and I put this on my personal todo list, and finally came around to do this. 

Back in that discussion you only cared that the `systemd-tmpfiles --purge` invocation in the debian packages could stay the same, and it can with this. You were fine if we tighten the rules on the drop-in files however.

If this is not opt-in like this patch makes it I am pretty sure it's not worth the effort to adding yet another flag.

File: man/systemd-tmpfiles.xml [hunk 2]
```diff
@@ -152,10 +152,11 @@
       <varlistentry>
         <term><option>--purge</option></term>
 
-        <listitem><para>If this option is passed, all files and directories marked for
-        <emphasis>creation</emphasis> by the <filename>tmpfiles.d/</filename> files specified on the command
-        line will be <emphasis>deleted</emphasis>. Specifically, this acts on all files and directories
-        marked with <varname>f</varname>, <varname>F</varname>, <varname>d</varname>, <varname>D</varname>,
+        <listitem><para>If this option is passed, all files and directories declared for
+        <emphasis>creation</emphasis> and marked wit the <literal>$</literal> character by the
```
Comment (daandemeyer) [inline man/systemd-tmpfiles.xml]: wit => with

**Review by DaanDeMeyer — APPROVED** [review]
I agree with this approach, It's better to be safe than sorry. Sure some extra work will be needed to add the specifier where required but at least like this when something goes wrong nobody can yell at us.

File: NEWS (hunk 1 above)
Comment (bluca) [reply NEWS:12]: actually there are some that are not for the future in the list below too, but we can rearrange later

**Comment by bluca** [comment]
> Oh man, I thought we had agreement to tighten the rules like this, after the fuck-up people ran into. We talked about this, and I put this on my personal todo list, and finally came around to do this.
> 
> Back in that discussion you only cared that the `systemd-tmpfiles --purge` invocation in the debian packages could stay the same, and it can with this. You were fine if we tighten the rules on the drop-in files however.
> 
> If this is not opt-in like this patch makes it I am pretty sure it's not worth the effort to adding yet another flag.

Sure, it was just another idea to make the transition less work with the same net effect, given the core idea is to safeguard a few top level locations, so in my view it would still be very well worth it. But never mind. Could you please then go through tmpfiles.d/ and add the new specifier where it makes sense to have it?

**Comment by bluca** [comment]
> but at least like this when something goes wrong nobody can yell at us.

<Press F for 'doubt'>

File: NEWS (hunk 1 above)
Comment (poettering) [reply NEWS:12]: ah, indeed, missed those. yeah, we should rearrange this to have two sections, one for stuff we broke and one of stuff we intend to break.

**Comment by bluca** [comment]
How about:

20-systemd-userdb.conf
portables.conf
systemd.conf
systemd-nspawn.conf
systemd-pstore.conf
systemd-tmp.conf

?

All of these contain very service-specific directories in /var/ or /etc/

**Comment by poettering** [comment]
> 20-systemd-userdb.conf

oversight, will fix.

> portables.conf
> systemd.conf
> systemd-nspawn.conf
> systemd-pstore.conf
> systemd-tmp.conf

These ones are either about components we really don't support removing. Or about directories which contain stuff not owned by us, hence I don't think we shiuld let this be cleaned up by --purge.

**Comment by bluca** [comment]
> > portables.conf
> > systemd.conf
> > systemd-nspawn.conf
> > systemd-pstore.conf
> > systemd-tmp.conf
> 
> These ones are either about components we really don't support removing. Or about directories which contain stuff not owned by us, hence I don't think we shiuld let this be cleaned up by --purge.

mmh /var/tmp/systemd-private /tmp/systemd-private /var/lib/systemd and /run/systemd is private data of our components, so sounds like things that should be removed? I guess things like /var/lib/machines and /var/lib/portables is a bit more borderline, but the rest seems like good candidates

**Comment by poettering** [comment]
> mmh /var/tmp/systemd-private /tmp/systemd-private /var/lib/systemd and /run/systemd is private data of our components, so sounds like things that should be removed? 

We do not support deinstallation of the service manager? hence doesn't really make sense to mark this for deletion?

> I guess things like /var/lib/machines and /var/lib/portables is a bit more borderline, but the rest seems like good candidates

these dirs contain user images, we shouldn't delete those.

but which else do you have in mind?

**Comment by bluca** [comment]
> > mmh /var/tmp/systemd-private /tmp/systemd-private /var/lib/systemd and /run/systemd is private data of our components, so sounds like things that should be removed?
> 
> We do not support deinstallation of the service manager? hence doesn't really make sense to mark this for deletion?

We do, it is possible to switch to a different init system and reboot into it, and then the systemd package can be removed, so I think the private dirs should be removed. There's no reason not to, if it cannot happen then the package cannot be removed and hence they will never be exercised. If they are, it's because the package has been purged, so the service manager is gone too.

> > I guess things like /var/lib/machines and /var/lib/portables is a bit more borderline, but the rest seems like good candidates
> 
> these dirs contain user images, we shouldn't delete those.
> 
> but which else do you have in mind?

A purge operation _does_ remove user data, so in principle it would be fair game, but I agree it's borderline and doesn't matter too much in the end

**Comment by poettering** [comment]
> We do, it is possible to switch to a different init system and reboot into it, and then the systemd package can be removed, so I think the private dirs should be removed. There's no reason not to, if it cannot happen then the package cannot be removed and hence they will never be exercised. If they are, it's because the package has been purged, so the service manager is gone too.
> 
> > >

These are directories in /run/. Sorry, but we do not support removal of PID 1 during runtime.

**Comment by bluca** [comment]
It's ok to skip /run, but please do add `$` to the private directories in /var as removing the package is supported after a reboot

File: NEWS (hunk 1 above)
Comment (YHNdnzj) [reply NEWS:12]: Ideally this can just be done in this PR?

**Review by bluca — APPROVED** [review]

File: NEWS (hunk 1 above)
Comment (poettering) [reply NEWS:12]: i started to rework NEWS in a bigger fashion in another git branch. would prefer to do that separately.

**Comment by poettering** [comment]
So, that that actually doesn't work. When you added the --purge logic you rightfully limited the effect on lines that actually *create* files. But these lines are "x" lines, i.e. exclusion, they do not create files, but exclude them.

We can certainly revisit this, but I think we shouldn't be more catholic than the pope on this,  for now at least, i.e. not suddenly purge *more* stuff than we did before this PR. If that makes sense.

or in other words, feel free to submit a PR later for this?

**Comment by poettering** [comment]
(CI found this out, now pushed another version that dropped this again)

**Comment by bluca** [comment]
Ah yes I had missed that those were `x` lines

**Review by yuwata — APPROVED** [review]

**Pull request merged**